feat(libp2p): add `SwarmBuilder`

mxinden opened the pull request:
## Description

Introduce the new `libp2p::SwarmBuilder`. Users should use the new `libp2p::SwarmBuilder` instead of the now deprecated `libp2p::swarm::SwarmBuilder`. See `libp2p::SwarmBuilder` docs on how to use the new builder.

Fixes #3657.
Fixes #3563.
Fixes #3179.

## Notes & open questions


<!--
Any notes, remarks or open questions you have to make about the PR which don't need to go into the final commit message.
-->

## Change checklist

<!-- Please add a Changelog entry in the appropriate crates and bump the crate versions if needed. See <https://github.com/libp2p/rust-libp2p/blob/master/docs/release.md#development-between-releases>-->

- [ ] I have performed a self-review of my own code
- [ ] I have made corresponding changes to the documentation
- [ ] I have added tests that prove my fix is effective or that my feature works
- [ ] A changelog entry has been made in the appropriate crates

mxinden committed:
feat(libp2p): add `TransportBuilder`

mxinden committed:
Introduce BuildableTransport abstraction

mxinden committed:
Feature flag

mxinden committed:
Add OtherTransportBuilder

mxinden committed:
Fix order

mxinden commented on libp2p/src/builder.rs:
```diff
@@ -0,0 +1,282 @@
+                .no_more_other_transports()
+                .without_dns()
+                .build();
+    }
```
`TransportBuilder` allows users to bring their own `Transport` implementations, e.g. `libp2p-quic`, even though they are unknown to `libp2p`. The only requirement is that they implement `Transport<Output = (PeerId, StreamMuxerBox)>`.

DougAnderson444 replied:
This is awesome!! 

I has also discovered that I had to do some shuffling with Transports if I wanted more than 2 fallbacks (dns, quic, and/or webrtc), like this:

https://github.com/DougAnderson444/libp2peasy/blob/ede291d7fc1554549a954c05927d1eb5051802fc/src/transport.rs#L64-L74 

Does this builder mean we wouldn't need to do that anymore? (If so, great, because `Either, Either...` types were a bit hard to keep track of)

The other idea I had was to "pick some sane defaults" for typical use, then give the option to pass in a custom Config overriding it but you've likely already considered that in here somewhere.

thomaseizinger replied:
> The other idea I had was to "pick some sane defaults" for typical use, then give the option to pass in a custom Config overriding it but you've likely already considered that in here somewhere. 

Whilst sounding appealing, this is quite the footgun due to how cargo features work.

You shouldn't have features that change functionality because any additional dependency in your tree could suddenly change behaviour in your code.

Thus, the better idea is to just make the builder as easy as possible but have the user explicitly say what they want.

thomaseizinger replied:
> I has also discovered that I had to do some shuffling with Transports if I wanted more than 2 fallbacks (dns, quic, and/or webrtc), like this:
> 
> https://github.com/DougAnderson444/libp2peasy/blob/ede291d7fc1554549a954c05927d1eb5051802fc/src/transport.rs#L64-L74 

Yes that is a known issue and will be addressed by this solution.

thomaseizinger reviewed (commented):
Great work Max!

Before you continue supporting more transports, I'd appreciate some exploration on how this can be turned into a `SwarmBuilder`.

Specifically, we should make it possible to pass the identity in there but also generate a new one and then pass it to the `Swarm` as well.

thomaseizinger commented on libp2p/src/builder.rs:
```diff
@@ -0,0 +1,282 @@
+            transport: libp2p_tcp::async_io::Transport::new(
+                libp2p_tcp::Config::new().nodelay(true),
+            )
+            .upgrade(libp2p_core::upgrade::Version::V1)
```
I think we should default to V1Lazy here in preparation of always making this the default.

thomaseizinger commented on libp2p/src/builder.rs:
```diff
@@ -0,0 +1,282 @@
+    pub fn with_relay(
+        self,
+        relay_transport: libp2p_relay::client::Transport,
+    ) -> OtherTransportBuilder<P, impl BuildableTransport> {
```
Once we make a `SwarmBuilder` out of this, it would be great if calling `with_relay` just changes the closure for `with_behaviour` from `(identity::Keypair) -> NB` to `(identity::Keypair, relay::Behaviour) -> NB`.

thomaseizinger commented on libp2p/src/builder.rs:
```diff
@@ -0,0 +1,282 @@
+    type E: Send + Sync + 'static;
+    type D: Send;
+    type U: Send;
+}
```
```suggestion
pub trait AuthenticatedMultiplexedTransport:
    Transport<
        Error = Self::E,
        Dial = Self::D,
        ListenerUpgrade = Self::U,
        Output = (libp2p_identity::PeerId, StreamMuxerBox),
    > + Send
    + Unpin
    + 'static
{
    type E: Send + Sync + 'static;
    type D: Send;
    type U: Send;
}
```

mxinden committed:
Introduce SwarmBuilder

- Allow providing or generating identity
- Use V1Lazy
- Make NetworkBehaviour constructor take relay Behaviour in case with_relay was called
- Rename to AuthenticatedMultiplexedTransport

thomaseizinger commented on libp2p/src/builder.rs:
```diff
@@ -0,0 +1,399 @@
+impl SwarmBuilder {
+    pub fn new() -> SwarmBuilder {
+        Self {}
+    }
```
Nit: Do we need this or should we just start directly with one of the choices, like the underlying runtime?

mxinden replied:
Matter of taste. I like having a conventional entry-point. I.e. without knowing anything about a `struct` I can depend on there being a `new` method. I don't feel strongly about it.

thomaseizinger commented on libp2p/src/builder.rs:
```diff
@@ -0,0 +1,399 @@
+        Self {}
+    }
+
+    pub fn with_new_identity(self) -> ProviderBuilder {
```
If somehow possible, I'd like to reduce the number of types that we need to expose for this.

mxinden replied:
Makes sense. That said, I don't know how. I could introduce a wrapper `struct` that delegates to a generic inner type. Though the delegation logic would introduce a lot of complexity.

thomaseizinger replied:
It does not necessarily reduce the number of types but what I've seen in other projects:

- One central `Builder` struct that has a generic state or phase
- The phases have fields themselves

See `rustls` for example: https://github.com/rustls/rustls/blob/main/rustls/src/builder.rs#L97

Something similar for us would be:

- `SwarmBuilder`
- states / phases called:
   - `WantsBaseTransport`
   - `WantsRelay`
   - `WantsOtherTransports`
- `impl` blocks for `SwarmBuilder<WantsRelay>` etc

mxinden replied:
Problem is that the "phases" differ in the number of type parameters they require:

``` rust
pub struct SwarmBuilder {}
pub struct ProviderBuilder {
pub struct TcpBuilder<P> {
pub struct RelayBuilder<P, T> {
pub struct OtherTransportBuilder<P, T, R> {
pub struct DnsBuilder<P, T, R> {
pub struct BehaviourBuilder<P, R> {
pub struct Builder<P, B> {
```

We could parameterize `SwarmBuilder` over the phase + all type parameters across all phases:

``` rust
pub struct SwarmBuilder<Phase, Provider, Transport, Relay, Behaviour>
```

thomaseizinger replied:
If the `SwarmBuilder` has a single type parameter for the phase, this should not matter? For example, `WantsBaseTransport` can have another type parameter for being generic over the supported runtime which can be passed on from a `WantsRuntime` phase, right?

mxinden replied:
If `WantsBaseTransport` is generic over a type parameter, then `SwarmBuilder` needs to provide that type parameter to `WantsBaseTransport` and thus needs to be generic over that same type parameter as well. Thereby I am not sure how your suggestion simplifies things. @thomaseizinger can you expand on your idea?

thomaseizinger replied:
I thought that we could do something like:

```
struct SwarmBuilder<TPhase> {

}

impl SwarmBuilder<WantsProvider> {
    fn with_tokio(self) -> SwarmBuilder<WantsTransport<Tokio>>;
}

impl SwarmBuilder<WantsTransport<Tokio>> {
   fn with_tcp()
}
```

Maybe I am missing something but I thought that this could work.

The phases would have to store all the data and the builder would only ever carry the phase, except for all non-generic stuff like identity.

mxinden replied:
> Maybe I am missing something but I thought that this could work.

Exactly the opposite. You are right. It is possible. Thank you for being meticulous here @thomaseizinger.



> It does not necessarily reduce the number of types

This still stands. But I think having a single root level type is easier for users to handle. In other words, I am in favor of the phase-style.

See d3ffdd4d2

thomaseizinger commented on libp2p/src/builder.rs:
```diff
@@ -0,0 +1,399 @@
+#[cfg(feature = "relay")]
+impl<P, T: AuthenticatedMultiplexedTransport> RelayBuilder<P, T> {
+    // TODO: This should be with_relay_client.
+    // TODO: Ideally one can configure it.
```
What is there to configure?

mxinden replied:
:+1: resolved.

thomaseizinger commented on libp2p/src/builder.rs:
```diff
@@ -0,0 +1,399 @@
+    }
+}
+
+pub struct NoRelayBehaviour;
```
Could just replace this with `()` maybe?

mxinden replied:
Yes, possible. I opted for a custom type to be specific. I find `NoRelayBehaviour` more descriptive than `()`.

thomaseizinger commented on libp2p/src/builder.rs:
```diff
@@ -0,0 +1,399 @@
+
+#[cfg(feature = "relay")]
+impl<P> BehaviourBuilder<P, libp2p_relay::client::Behaviour> {
+    // TODO: The close should provide the relay transport in case the user used with_relay.
```
Resolved?

thomaseizinger commented on libp2p/src/builder.rs:
```diff
@@ -0,0 +1,399 @@
+        Self {}
+    }
+
+    pub fn with_new_identity(self) -> ProviderBuilder {
```
```suggestion
    pub fn with_ephemeral_identity(self) -> ProviderBuilder {
```

thomaseizinger commented on libp2p/src/builder.rs:
```diff
@@ -0,0 +1,399 @@
+            .with_tokio()
+            .with_tcp()
+            .without_relay()
+            .with_other_transport(libp2p_core::transport::dummy::DummyTransport::new())
```
QUIC and WebRTC both will need access to the keypair so this should be a closure I'd say.

mxinden replied:
:+1: done in https://github.com/libp2p/rust-libp2p/pull/4120/commits/343ba2d2729c4349544c4b027f8a2d6cc29e398b.

thomaseizinger reviewed (commented):
This is looking very nice already. I'd say that is worth a blog-post once we ship it :)

Some good content for TWIR!

mxinden committed:
Add with_custom_tcp

mxinden committed:
Have with_other_transport take keypair closure

mxinden committed:
Fully qualify libp2p_swarm::NetworkBehaviour

mxinden committed:
Use SwwarmBuilder in DCUtR example

mxinden committed:
Remove resolved TODO

mxinden committed:
Remove custom TCP option

I suggest shipping a minimal Builder at first. Users that need advanced features can use the
`with_other_transport` method or build a `Transport` without the builder.

mxinden committed:
Add without_tcp option

mxinden committed:
feat(core): add upgrade::Map

mxinden committed:
feat(swarm/builder): make authentication configurable

- Make tls and noise configurable.
- Prevent accidental behavior change through feature change. E.g. enabling "noise" feature should
  not make ones transport automatically offer noise. Instead one needs to explicitly call
  `with_noise` and `without_noise`.

mxinden committed:
Feature flag for relay

mxinden committed:
refactor(examples/ping): use `SwarmBuilder`

mxinden committed:
Rename to without_any_other_transports

mxinden commented on core/src/upgrade/map.rs:
Needed for `TransportBuilder`, more specifically in order to use both `libp2p-noise` and `libp2p-tls` on a `libp2p-tcp` or `libp2p-relay` transport.

I am happy to split this change out into a separate pull request.

thomaseizinger replied:
Can we make it an implementation detail of the `TransportBuilder` instead of increasing the public API?

mxinden replied:
In my eyes it is worth extending `libp2p-core`'s public API surface, especially as it is well isolated. Say e.g. someone wants to combine both `libp2p-noise` and `libp2p-tls` but does not want to use `SwarmBuilder`. In such case they would need `upgrade::Map`.

thomaseizinger replied:
I disagree, `libp2p-core` needs to go away eventually and adding to its API now doesn't help with that.

> but does not want to use `SwarmBuilder`

Then they will have to write some extra code or copy `upgrade::Map` to their project. I expect this new `SwarmBuilder` to handle the majority if not all usecases known to us.

DRY is the root of all evil in regards to dependencies and in this case, making it an implementation detail isn't even duplicating anything. I am open to discuss making it a public component once WE have a 2nd usecase for it :)

thomaseizinger replied:
Do you feel strongly about exposing `Map` in the public API? I don't think we are doing ourselves a favor with it ..

thomaseizinger commented on libp2p/src/builder.rs:
```diff
@@ -59,13 +153,40 @@ impl TcpBuilder<AsyncStd> {
             transport: libp2p_tcp::tokio::Transport::new(Default::default())
                 .upgrade(libp2p_core::upgrade::Version::V1Lazy)
-                .authenticate(libp2p_noise::Config::new(&self.keypair).unwrap())
+                // TODO: Handle unwrap?
```
I'd say we just return a `Result` from this builder function. It should still be pretty ergonomic together with `?` for the user.

mxinden replied:
Done :+1:

thomaseizinger commented on libp2p/src/builder.rs:
```diff
@@ -373,12 +630,20 @@ mod tests {
+            .with_noise()
             .without_relay()
             .no_more_other_transports()
             .with_dns()
```
It would probably lead to a bit of duplication but can we add a `with_behaviour` function to the builder after `with_noise` that simply calls these three functions?

mxinden replied:
Added the shortcut. Good idea.

mxinden replied:
Done

thomaseizinger commented on examples/dcutr/src/main.rs:
```diff
@@ -155,21 +122,30 @@ fn main() -> Result<(), Box<dyn Error>> {
+            )),
+            dcutr: dcutr::Behaviour::new(keypair.public().to_peer_id()),
+        })
+        .build();
```
Do we need `.build` or can that not be implied by `with_behaviour`? Perhaps `.build` should take the closure? I think it is worth optimizing that users don't need to call "unnecessary" functions here.

mxinden replied:
I am undecided. I like a `Builder` having a final `build` method. Including `build` in `with_behaviour` does not make it explicit that it is the final step. Including `with_behaviour` in `build` does not make it explicit that it wants a `NetworkBehaviour`.

I am fine with as is. In case you feel strongly about it, can you think of any other method names?

thomaseizinger replied:
> Including `with_behaviour` in `build` does not make it explicit that it wants a `NetworkBehaviour`.

It would be explicit because it takes an argument. Looking at it just as a builder-pattern, I think it is okay to include all required parameter in the `build` function. Traditionally, you would require them in the builder constructor but we cannot do that because the behaviour may need the identity key etc. So moving that to the `build` function is the next best thing IMO.

mxinden replied:
`with_existing_identity` takes a `Keypair`, `with_tcp_config` takes a `TcpConfig`, `with_other_transport` takes another `Transport`. I find `build` taking a `NetworkBehaviour` inconsistent.

thomaseizinger replied:
Would `build_with_behaviour` be too verbose?

mxinden replied:
In my eyes it combines two unrelated actions. `with_behaviour` is like `with_tcp` and `with_other_transport` in regards to providing some input. `build` is a final step combining all of the previous inputs. It does not make sense to me to combine `build` with one of the input actions (i.e. `with_behaviour`).

thomaseizinger commented on examples/dcutr/src/main.rs:
```diff
@@ -155,21 +122,30 @@ fn main() -> Result<(), Box<dyn Error>> {
+                .map(|(peer_id, muxer), _| (peer_id, StreamMuxerBox::new(muxer)))
+        })
+        .no_more_other_transports()
+        .with_dns()
```
Similar to another comment, can `with_dns` not impl `no_more_other_transports`?

mxinden replied:
:+1: Done in a6cf98f50.

mxinden committed:
Merge branch 'master' of https://github.com/libp2p/rust-libp2p into transport-builder

mxinden committed:
feat: websocket support

mxinden committed:
Use macros for de-duplication

mxinden committed:
Fix tests

mxinden committed:
Allow configuring tcp

mxinden committed:
Add shortcuts

mxinden committed:
Update autonat examples

mxinden committed:
Update chat examples

mxinden committed:
Fix dcutr example

mxinden committed:
Add shortcuts

mxinden committed:
Update key value store example

mxinden committed:
Update file sharing example

thomaseizinger commented on examples/autonat/src/bin/autonat_client.rs:
```diff
@@ -49,20 +49,17 @@ async fn main() -> Result<(), Box<dyn Error>> {
-    let local_key = identity::Keypair::generate_ed25519();
-    let local_peer_id = PeerId::from(local_key.public());
-    println!("Local peer id: {local_peer_id:?}");
+    let mut swarm = libp2p::builder::SwarmBuilder::new()
```
```suggestion
    let mut swarm = libp2p::SwarmBuilder::new()
```

What do you think of re-exporting this?

mxinden replied:
:+1: https://github.com/libp2p/rust-libp2p/pull/4120/commits/cb67678abdb92be7f9f68dbff66968224820a25c

thomaseizinger commented on examples/chat-example/src/main.rs:
```diff
@@ -44,58 +41,54 @@ struct MyBehaviour {
         })
-        .boxed();
+        .without_any_other_transports()
+        .with_behaviour(|key| {
```
I think `with_behaviour` will need to be fallible?

Ideally though, we can design it such that the common case where it is not fallible doesn't "pay" for that by having to `Ok` wrap everything.

mxinden replied:
I played around with making `with_behaviour` fallible.

A type parameter `E` is an option:

```
    pub fn with_behaviour<B, E>(
        self,
        mut constructor: impl FnMut(&libp2p_identity::Keypair, libp2p_relay::client::Behaviour) -> Result<B, E>,
    ) -> Result<Builder<P, B>, E> {
```

Though even in the case where `constructor` explicitly returns an error, rustc is not able to infer the error type.

```
error[E0282]: type annotations needed
  --> examples/chat-example/src/main.rs:78:13
   |
78 |             Ok(MyBehaviour { gossipsub, mdns })
   |             ^^ cannot infer type of the type parameter `E` declared on the enum `Result`
   |
help: consider specifying the generic arguments
   |
78 |             Ok::<MyBehaviour, E>(MyBehaviour { gossipsub, mdns })
   |               ++++++++++++++++++

```

Unfortunately one can not provide a defautl value for the `E` type parameter.

```
error: defaults for type parameters are only allowed in `struct`, `enum`, `type`, or `trait` definitions
   --> libp2p/src/builder.rs:826:30
    |
826 |     pub fn with_behaviour<B, E = void::Void>(
    |                              ^^^^^^^^^^^^^^
    |
    = warning: this was previously accepted by the compiler but is being phased out; it will become a hard error in a future release!
    = note: for more information, see issue #36887 <https://github.com/rust-lang/rust/issues/36887>
    = note: `#[deny(invalid_type_param_default)]` on by default
```

What does work is using `Box<dyn std::error::Error>`.

``` rust
    pub fn with_behaviour<B>(
        self,
        mut constructor: impl FnMut(&libp2p_identity::Keypair, libp2p_relay::client::Behaviour) -> Result<B, Box<dyn std::error::Error>>,
    ) -> Result<Builder<P, B>, Box<dyn std::error::Error>> {
```

@thomaseizinger let me know if you are fine with `Box<dyn std::error::Error>`.

---

> Ideally though, we can design it such that the common case where it is not fallible doesn't "pay" for that by having to `Ok` wrap everything.

I see the motivation. I don't see how to do it.

I tried type overloading:

``` rust
    pub fn with_behaviour<B>(
        self,
        mut constructor: impl FnMut(&libp2p_identity::Keypair, libp2p_relay::client::Behaviour) -> B,
    ) -> Result<Builder<P, B>, Box<dyn std::error::Error>> {
        self.with_behaviour(|key| Ok(constructor(key)))
    }
```

but that is not legal:

```
error[E0592]: duplicate definitions with name `with_behaviour`
   --> libp2p/src/builder.rs:832:9
    |
826 | /     pub fn with_behaviour<B>(
827 | |         self,
828 | |         mut constructor: impl FnMut(&libp2p_identity::Keypair, libp2p_relay::client::Behaviour) -> B,
829 | |     ) -> Result<Builder<P, B>, Box<dyn std::error::Error>> {
    | |__________________________________________________________- other definition for `with_behaviour`
...
832 | /         pub fn with_behaviour<B>(
833 | |             self,
834 | |             mut constructor: impl FnMut(&libp2p_identity::Keypair, libp2p_relay::client::Behaviour) -> Result<B, Box<dyn std::error::Erro...
835 | |         ) -> Result<Builder<P, B>, Box<dyn std::error::Error>> {
    | |______________________________________________________________^ duplicate definitions for `with_behaviour`
```

thomaseizinger replied:
I pushed a commit that I think fixes this reasonably well.

mxinden replied:
:rocket: thank you @thomaseizinger!

thomaseizinger commented on examples/dcutr/src/main.rs:
```diff
@@ -155,21 +122,34 @@ fn main() -> Result<(), Box<dyn Error>> {
-        Err(_) => SwarmBuilder::without_executor(transport, behaviour, local_peer_id),
-    }
-    .build();
+    info!("Local peer id: {:?}", swarm.local_peer_id());
```
Worth considering adding a generic log to the boot of a `Swarm` I think? In separate PR though :)

mxinden replied:
Good idea. Tracked in https://github.com/libp2p/rust-libp2p/issues/4394.

thomaseizinger reviewed (commented):
One of the biggest QoL improvements in a while! Looking forward to this!

mxinden committed:
Update metrics example

mxinden committed:
Update libp2p identify example

mxinden committed:
Update ipfs-kad example

mxinden committed:
Update ipfs-private example

mxinden committed:
Update relay server

mxinden committed:
Update rendezvous examples

mxinden committed:
Export as libp2p::SwarmBuilder

mxinden committed:
Add shortcuts to skip without_any_other_transports

mxinden committed:
Error handling

mxinden committed:
Make with_behaviour fallible

mxinden committed:
Add with_quic

mxinden committed:
Merge branch 'master' of https://github.com/libp2p/rust-libp2p into transport-builder

mxinden committed:
Use with_quic

thomaseizinger commented on libp2p/src/builder.rs:
```diff
@@ -0,0 +1,1203 @@
+    #[error("Noise")]
+    #[cfg(feature = "noise")]
+    Noise(#[from] libp2p_noise::Error),
+}
```
This is a semver-hazard because activating a feature can suddenly make an exhaustive match non-exhaustive. We should encapsulate this using a struct.

mxinden replied:
What do you think of 46a8bd982?

Would appreciate input on how to circumvent all the `from` and `into` calls @thomaseizinger.

thomaseizinger replied:
Using `?` will call `From` for you so you could implement `From<libp2p_noise::Error> for AuthenticationError` and use `?` without having to call `map_err`.

mxinden replied:
Related fix: 3e34a6581

thomaseizinger commented on libp2p/src/builder.rs:
```diff
@@ -0,0 +1,1203 @@
+
+    pub fn with_behaviour<B>(
+        self,
+        constructor: impl FnMut(&libp2p_identity::Keypair) -> Result<B, Box<dyn std::error::Error>>,
```
Can we avoid the `Ok` wrapping if we do something like:


```suggestion
        constructor: impl FnMut(&libp2p_identity::Keypair) -> O,
```
and

```rust
where O: Into<Result<B, E>>
```

And then provide an implementation like: 

```rust
impl Into<Result<B, Infallible>> for B where B: NetworkBehaviour { }
```

thomaseizinger commented on libp2p/src/builder.rs:
```diff
@@ -0,0 +1,1203 @@
+            .without_dns()
+            .without_websocket()
+            .with_behaviour(constructor)
+    }
```
I imagine we could create some macros that de-duplicate the function definition but require you to pass the body. Then we could at least somewhat be sure that we don't have any typos or differences in method signatures.

mxinden replied:
I experimented with this. I don't think the macro actually improves maintenance nor readability due to the indirection and additional complexity.

https://github.com/mxinden/rust-libp2p/commit/3bdc34ff3cba33403e05832bff3eaa571e1c67b7

I have not included the commit on this branch.

thomaseizinger replied:
Fair!

thomaseizinger commented on libp2p/src/builder.rs:
```diff
@@ -0,0 +1,1203 @@
+    }
+
+    // TODO: Not the ideal name.
+    pub fn without_any_other_transports(
```
Given that we have the short-cuts, does it need to be public?

thomaseizinger replied:
I think the same applies for all other `without_` functions.

mxinden replied:
Good idea. Done via faecb8bcc.

thomaseizinger replied:
Can you expand on your commit message? Why is it not possible for `without_` functions that return a `Result`? Shouldn't we be able to always have a shortcut function to the next builder?

mxinden replied:
_For sure. I actually just now tried to ignorantly tackle this again and made some progress._

We can actually make `without_noise` not public on `TcpNoisePhase`. We create the `libp2p_tls::Config` in `with_tls` in `TcpTlsPhase` already and can thus already return `AuthenticationError` in `with_tls`.

4ad2a0e42

The same does not work for `WebsocketPhase`. When the user calls `with_tls` we build the `libp2p_tls::Config`. But we don't know yet whether the user will call `with_noise` or `without_noise`. Thus we can not yet build the additional DNS transport required for websocket, where building the DNS transport can error. Now without the `without_noise` method, we would have to return the errors in the shortcut, i.e. `with_behaviour`. In other words we would have to return an error in `with_behaviour` that arose in `without_noise`. I don't think that is ideal.

I don't think the ergonomics of Websocket creation are as critical as the ergonomics around TCP. Thus I am fine with the progress just made.

@thomaseizinger this is difficult to explain and understand. Let me know if you want me to expand further.

thomaseizinger replied:
Yeah this makes sense and I agree with the conclusion, thank you!

thomaseizinger commented on libp2p/src/builder.rs:
```diff
@@ -0,0 +1,1203 @@
+            self.behaviour,
+            self.keypair.public().to_peer_id(),
+        )
+        .build()
```
Long-term, I'd imagine that we ditch the other `SwarmBuilder` and only provide a single API for constructing a `Swarm`.

thomaseizinger commented on libp2p/src/lib.rs:
```diff
@@ -142,6 +142,9 @@ pub use libp2p_webtransport_websys as webtransport_websys;
 pub use libp2p_yamux as yamux;
 
+mod builder;
+pub use builder::SwarmBuilder;
```
Let's group this with the others.

mxinden replied:
I don't understand @thomaseizinger. You want me to remove the newline to `mod transport_ext`?

``` rust
mod builder;
pub use builder::SwarmBuilder;

mod transport_ext;

pub mod bandwidth;

#[cfg(doc)]
pub mod tutorials;
```

thomaseizinger replied:
We have `pub use` statements further down. The `pub use builder::SwarmBuilder` statement should be there IMO.

mxinden replied:
Moved with c289ccc3a.

mxinden committed:
Introduce phases

thomaseizinger commented on libp2p/src/builder.rs:
```diff
@@ -0,0 +1,1301 @@
+// TODO: Rename runtime to provider
+// TODO: Should we have a timeout on transport?
+// TODO: Be able to address `SwarmBuilder` configuration methods.
```
I'd suggest that we copy all those over to this `SwarmBuilder`. You could introduce a phase that carries the current `libp2p_swarm::SwarmBuilder` inside and set the options on that one. Later, when we remove the `libp2p_swarm::SwarmBuilder` we can set track those values directly.

mxinden replied:
I had a similar idea. How about we introduce a `Config` in `libp2p-swarm`, i.e. `libp2p_swarm::Config`.

- That would be consistent with all other `libp2p-*` crates. (Though some still use the `CratenameConfig` syntax.)
- We can introduce this in a non-breaking way, i.e. have the existing `SwarmBuilder` in `libp2p-swarm` use the new `libp2p_swarm::Config` internally.
- In the builder introduced in this pull request, we can have a `with_swarm_config` option, just like we have a `with_tcp_config`.

Thoughts?

thomaseizinger replied:
Great idea!

mxinden replied:
Drafted in 2b687276f.

thomaseizinger commented on libp2p/src/builder.rs:
```diff
@@ -0,0 +1,1301 @@
+            phantom: PhantomData,
+            phase: RelayPhase {
+                transport: libp2p_core::transport::dummy::DummyTransport::new(),
+            },
```
To use a relay, one needs some kind of base transport. Why should it be legal to go to the relay phase without first saying `with_tcp` or `with_quic`?

mxinden replied:
> Why should it be legal to go to the relay phase without first saying `with_tcp` or `with_quic`?

`with_tcp`, `with_quic` or `with_other_transport`

Fair. Let me see what I can do. Thus far I have shied away from its complexity.

mxinden replied:
I don't see a reasonable way around this. One could make `with_quic` (...) generic over the previous transport, where a previous `DummyTransport` would lead to it being dropped in `with_quic` and `build` to only be available when the final transport is not a `DummyTransport`. I don't think it is worth the complexity this would introduce.

thomaseizinger commented on libp2p/src/builder.rs:
```diff
@@ -0,0 +1,1301 @@
+// TODO: Be able to address `SwarmBuilder` configuration methods.
+// TODO: Consider moving with_relay after with_other_transport.
+// TODO: Consider making with_other_transport fallible.
+
```
More TODOs:

- Can this handle custom authentication algorithms? Do we need to be able to handle that?
- Can this handle custom base transports, like https://github.com/umgefahren/libp2p-tor? Should it?

I think as a first step, we don't need to handle these cases. In the long run, looking at https://github.com/libp2p/rust-libp2p/issues/3271, it would be great if we could make all of the upgrade stuff an implementation detail of this `SwarmBuilder` and only expose the bare `Transport` trait. In other words, pack all usecases into this `SwarmBuilder`.

Do you think that is a good direction?

mxinden replied:
> Can this handle custom authentication algorithms? Do we need to be able to handle that?

I don't think so. You can do so via the `with_other_transport` mechanism.

> Can this handle custom base transports, like https://github.com/umgefahren/libp2p-tor? Should it?

Yes, see `with_other_transport`.

> I think as a first step, we don't need to handle these cases. In the long run, looking at #3271, it would be great if we could make all of the upgrade stuff an implementation detail of this `SwarmBuilder` and only expose the bare `Transport` trait. In other words, pack all usecases into this `SwarmBuilder`.
> 
> Do you think that is a good direction?

> Make the easy things easy and the hard things possible.

My number one goal of the new `SwarmBuilder` is the former, i.e. "easy things easy". I suggest we focus on that in this pull request.

As long as making the "hard things possible" doesn't come at the expense of the easy things, I am fine with the direction you are suggesting above.

With more and more `Transport`'s being upgraded out-of-the-box (e.g. `libp2p-quic`, `libp2p-webrtc`, ...), dedicated upgrade mechanisms (e.g. `Transport::upgrade`) become less relevant.

mxinden committed:
Make without_ methods private whereever possible

In the case where the without_ method returns a Result this is not possible.

mxinden committed:
Change phase order

- Move relay to the back. It doesn't need to come that early. Removes the need to drag the relay
behaviour through all phases.
- Move other transport closer to tcp and quic to benefit from dns.

mxinden committed:
Rename runtime to provider

mxinden committed:
Merge branch 'master' of https://github.com/libp2p/rust-libp2p into transport-builder

thomaseizinger committed:
Avoid need for `Ok` wrapping behaviour when closure is infallible

mxinden committed:
Apply global timeout of 10s

mxinden committed:
feat(swarm): add SwarmConfig

thomaseizinger commented on swarm/src/lib.rs:
```diff
@@ -356,6 +356,26 @@ impl<TBehaviour> Swarm<TBehaviour>
 {
+    /// Creates a new [`Swarm`] from the given [`Transport`], [`NetworkBehaviour`], [`PeerId`] and
+    /// [`SwarmConfig`].
+    pub fn new_with_config(
```
Can this just be `new` instead and deprecate all other constructors?

mxinden replied:
fe794518f

thomaseizinger commented on swarm/src/lib.rs:
```diff
@@ -1343,14 +1363,154 @@ impl<'a> PollParameters for SwarmPollParameters<'a> {
     }
 }
 
+pub struct SwarmConfig {
```
```suggestion
pub struct Config {
```

thomaseizinger replied:
Let's follow #2217 here.

mxinden committed:
Make `without_noise` on `TcpNoisePhase` private

mxinden committed:
Move pub use down

mxinden committed:
Deprecate development_transport

mxinden committed:
Wrap errors

mxinden committed:
Make RelayNoisePhase without_noise private

mxinden committed:
Add example showcasing all features

mxinden committed:
Merge branch 'master' of https://github.com/libp2p/rust-libp2p into transport-builder

mxinden committed:
Add with_quic_config

mxinden committed:
refactor(misc/server): use new SwarmBuilder

mxinden committed:
Use FnOnce instead of FnMut

mxinden committed:
Feature gate most from wasm32

mxinden committed:
Fix clippy

mxinden committed:
More feature flags

mxinden committed:
More feature flags

mxinden committed:
More feature flags

mxinden committed:
Fix feature flag

mxinden committed:
Clippy

mxinden committed:
Clippy

mxinden committed:
Merge branch 'master' of https://github.com/libp2p/rust-libp2p into transport-builder

mxinden committed:
Make with_other_transport fallible

mxinden committed:
Use SwarmBuilder in interop-tests

mxinden committed:
Add WASM support

mxinden committed:
Simplify with_wasm_bindgen

mxinden committed:
Remove wasm TODO

mxinden committed:
interop: remove mplex support

mxinden commented on interop-tests/native-ping-version.json:
```diff
@@ -12,7 +12,6 @@
     "noise"
   ],
   "muxers": [
-    "mplex",
```
Calling out this change explicitly. Supporting mplex requires some additional effort, given that the `libp2p::SwarmBuilder` only supports yamux. I don't think this additional effort is worth it. Any objections?

Additional context: https://github.com/libp2p/test-plans/pull/263

thomaseizinger replied:
I'd like there to be the possibility to have external multiplexers. If we want to be a modular network stack, we IMO cannot not offer the ability to plug in your own multiplexer. I know that it is possible via `with_other_transport` but I want to eventually remove the other APIs and thus would like this builder to support this from the beginning.

mxinden replied:
Re-introduced mplex in 9ee251fbf.

mxinden committed:
Update ping tutorial

mxinden committed:
Add with_bandwidth_logging

mxinden committed:
Shortcut for with_bandwidth_logging

mxinden committed:
Fix minor CI failures

mxinden committed:
Feature flag any(tls, noise)

mxinden committed:
Use SwarmConfig

mxinden committed:
Typo

mxinden committed:
Don't build swarm

mxinden committed:
fmt

mxinden committed:
Bump versions and add changelog entries

mxinden committed:
Fix relay tests

mxinden committed:
Move upgrade::Map into builder/map.rs

mxinden committed:
Bail on mplex

mxinden commented on interop-tests/src/arch.rs:
```diff
@@ -47,87 +35,97 @@ pub(crate) mod native {
+                    .with_tokio()
+                    .with_quic()
+                    .with_behaviour(behaviour_constructor)?
+                    .build();
```
Note that this is using a 10s instead of the previous 5s timeout now.

I argue that 10s is good enough for almost all cases. Thus I only want to make it configurable in case there is demand.

mxinden commented on swarm/src/lib.rs:
```diff
@@ -1343,14 +1363,146 @@ impl<'a> PollParameters for SwarmPollParameters<'a> {
+        Self {
+            pool_config: PoolConfig::new(None),
+        }
+    }
```
I find this to be a valid question. Should we carry the legacy `without_executor` over to the new `SwarmConfig`? In case there is demand, we can still introduce it in the future.

thomaseizinger replied:
Yeah I am okay to remove it.

mxinden replied:
4853caf88

mxinden committed:
Remove obsolete TODO

mxinden committed:
Feature flag map module

mxinden committed:
More feature flags

mxinden committed:
Remove use for map

mxinden committed:
Make Upgrade::new pub(crate)

mxinden committed:
Clippy

thomaseizinger commented on examples/ipfs-private/src/main.rs:
```diff
@@ -118,76 +95,72 @@ async fn main() -> Result<(), Box<dyn Error>> {
 
     // We create a custom network behaviour that combines gossipsub, ping and identify.
     #[derive(NetworkBehaviour)]
-    #[behaviour(to_swarm = "MyBehaviourEvent")]
```
Mind sending a separate PR for this?

mxinden replied:
:+1: https://github.com/libp2p/rust-libp2p/pull/4580

thomaseizinger commented on examples/ipfs-private/src/main.rs:
```diff
@@ -118,76 +95,72 @@ async fn main() -> Result<(), Box<dyn Error>> {
+                .authenticate(noise_config)
+                .multiplex(yamux_config)
+                .map(|(peer_id, muxer), _| (peer_id, StreamMuxerBox::new(muxer)))
+        })?
```
At the moment `with_other_transport` requires us to produce an authenticated and multiplexed transport. There are several usecases, including this one, where you only want to swap out the "base"(?) transport and still configure a security protocol and multiplexer on top.

Can we introduce something like `with_base_transport`? The `TcpPhase` would have to be renamed to `BaseTransportPhase` and represent any reliable, single-stream transport.

mxinden replied:
While _basetransport_ is easy to grasp for us maintainers, familiar with the upgrade flow of TCP, I don't think the same is true for regular users. I.e. I don't think they would know that TCP is a basetransport whereas QUIC is a full transport. Thus I don't think they would know that they have to call `with_base_transport` with a TCP transport.

Next not all basetransports are equal. While TCP can be placed anywhere within the hierarchy, websocket has to come last, wrapping a DNS and a TCP transport. Thus I don't think the `with_base_transport` abstraction is solid.

Does that make sense @thomaseizinger?

thomaseizinger replied:
Sounds good. This comment predates our synchronous conversation about this where I think we agreed that we are going to try something like:

```rust
fn with_tcp(self, security_protocols: [TSec; NUM_SEC], muxer_protocols: [TMux; NUM_MUX])
```

When typing out this API sketch, I noticed that arrays are maybe not such a good idea here because they aren't heterogeneous. Perhaps we should do something like:

```rust
fn with_tcp(self, security_protocols: impl IntoSec, muxer_protocols: impl IntoMuxer);

trait IntoSec {
	fn into_sec(self) -> Upgrade;
}

impl<T1> IntoSec for T1 { }
impl<T1, T2> IntoSec for (T1, T2) { }
// etc
```

That would allow the user to call it with tuples of different types. Otherwise we might have to introduce a boxed upgrade and calling the API isn't quite as convenient. For now, I'd suggest we implement it for up to 2 elements because that is how many muxers & security protocols we have. If more appear, users can always open an issue.

thomaseizinger commented on examples/rendezvous/src/bin/rzv-register.rs:
```diff
@@ -20,37 +20,33 @@
-        tcp::tokio::Transport::default()
-            .upgrade(Version::V1Lazy)
-            .authenticate(noise::Config::new(&key_pair).unwrap())
-            .multiplex(yamux::Config::default())
```
Why is `yamux` not part of the configuration?

mxinden replied:
I consider `yamux` the only valid muxer on top of TCP. In addition I consider the `with_tcp` flow to be an opinionated way to run TCP.

I chose not to have a `with_yamux` to reduce the complexity of the builder. Users that want to configure Yamux or any other muxer can use `with_other_transport`.

thomaseizinger commented on interop-tests/src/arch.rs:
```diff
@@ -47,87 +35,97 @@ pub(crate) mod native {
+            Muxer::Yamux => (),
+            Muxer::Mplex => {
+                bail!("Only Yamux is supported, not Mplex")
+            }
```
Just remove the enum variant entirely? Why keep parsing it if it is not supported. Plus, can you land this as a separate PR please?

mxinden replied:
I reintroduced mplex in 9ee251fbf

thomaseizinger commented on interop-tests/src/lib.rs:
```diff
@@ -255,3 +240,15 @@ where
         .map_err(Into::into)
 }
+
+pub(crate) fn build_behaviour(key: &Keypair) -> Behaviour {
```
Why isn't this just a closure?

mxinden replied:
I find functions to have a lower mental overhead compared to closures. With an `fn` it is clear that they do not capture anything. If you feel strongly about it, I am happy to use a closure instead.

thomaseizinger commented on libp2p/CHANGELOG.md:
```diff
@@ -1,3 +1,12 @@
+## 0.52.4
```
```suggestion
## 0.52.4 - unreleased
```

thomaseizinger commented on libp2p/CHANGELOG.md:
```diff
@@ -1,3 +1,12 @@
+
+- Introduce the new `libp2p::SwarmBuilder`.
+  Users should use the new `libp2p::SwarmBuilder` instead of the now deprecated `libp2p::swarm::SwarmBuilder`.
+  See `libp2p::SwarmBuilder` docs on how to use the new builder.
```
This is kind of implied, no?

mxinden replied:
I want to particularly stress this instance, as the docs show the order on which to call the builder methods. Feel strongly about it @thomaseizinger?

thomaseizinger commented on libp2p/CHANGELOG.md:
```diff
@@ -1,3 +1,12 @@
+## 0.52.4
+
+- Introduce the new `libp2p::SwarmBuilder`.
+  Users should use the new `libp2p::SwarmBuilder` instead of the now deprecated `libp2p::swarm::SwarmBuilder`.
```
```suggestion
- Introduce a new `libp2p::SwarmBuilder` in favor of the now deprecated `libp2p::swarm::SwarmBuilder`.
```

mxinden replied:
80b8657f3

thomaseizinger commented on libp2p/src/builder/map.rs:
Perhaps rename this to `map_upgrade`.

thomaseizinger commented on libp2p/src/lib.rs:
```diff
@@ -142,13 +142,16 @@ pub use libp2p_webtransport_websys as webtransport_websys;
 pub use libp2p_yamux as yamux;
 
+mod builder;
+
```
```suggestion
```

thomaseizinger commented on swarm/CHANGELOG.md:
```diff
@@ -1,3 +1,12 @@
+## 0.43.4 - unreleased
+
+- Deprecate old `libp2p::swarm::SwarmBuilder`.
```
```suggestion
- Deprecate `libp2p::swarm::SwarmBuilder`.
```

mxinden replied:
63633dcfc

thomaseizinger commented on libp2p/src/builder.rs:
```diff
@@ -0,0 +1,1795 @@
+}
+
+impl<Provider> SwarmBuilder<Provider, TcpPhase> {
+    // TODO: This would allow one to build a faulty transport.
```
Can this be removed now?

mxinden replied:
The underlying issue still prevails. That said, I don't think it is worth solving. Thus I am resolving here.

thomaseizinger commented on libp2p/src/builder.rs:
```diff
@@ -0,0 +1,1795 @@
+    phase: Phase,
+}
+
+pub struct InitialPhase {}
```
In rustls, these phases are called `WantsXYZ`. Would that be applicable here too? I am gonna make some suggestions.

mxinden replied:
I prefer the term `phase` as it is a noun and not a verb. Steps are declared as `struct`s and not as methods, thus in my eyes a noun makes more sense than a verb. @thomaseizinger let me know if you feel strongly about it.

thomaseizinger commented on libp2p/src/builder.rs:
```diff
@@ -0,0 +1,1795 @@
+    phase: Phase,
+}
+
+pub struct InitialPhase {}
```
```suggestion
pub struct WantsIdentity {}
```

thomaseizinger commented on libp2p/src/builder.rs:
```diff
@@ -0,0 +1,1795 @@
+    }
+}
+
+pub struct ProviderPhase {}
```
```suggestion
pub struct WantsRuntime {}
```

thomaseizinger commented on libp2p/src/builder.rs:
```diff
@@ -0,0 +1,1795 @@
+    }
+}
+
+pub struct TcpPhase {}
```
```suggestion
pub struct WantsBaseTransport {}
```

thomaseizinger commented on libp2p/src/builder.rs:
```diff
@@ -0,0 +1,1795 @@
+    feature = "tcp",
+    any(feature = "tls", feature = "noise")
+))]
+pub struct TcpTlsPhase {
```
```suggestion
pub struct WantsEncryption {
```

thomaseizinger commented on libp2p/src/builder.rs:
```diff
@@ -0,0 +1,1795 @@
+#[cfg(feature = "relay")]
+impl<Provider, T> SwarmBuilder<Provider, RelayTlsPhase<T>> {
+    #[cfg(all(not(target_arch = "wasm32"), feature = "tls"))]
+    pub fn with_tls(
```
I find it a bit confusing that this function is also called `with_tls` but it describes the encryption that is used OVER a relayed connection. Should we rename this to `with_relayed_tls` or something like that?

mxinden replied:
Solved with the recent refactor to use `with_tcp(security_protocol, muxer)`.

thomaseizinger commented on libp2p/src/builder.rs:
```diff
@@ -0,0 +1,1795 @@
+    }
+}
+
+// Shortcuts
```
This file is very large. Can we split this into multiple modules? For example, I think all the shortcuts are not very important so they could all go in a module. All macros could go into a module too.

thomaseizinger commented on libp2p/src/builder.rs:
```diff
@@ -0,0 +1,1795 @@
+    ) -> (
+        SwarmBuilder<Provider, BehaviourPhase<impl AuthenticatedMultiplexedTransport, R>>,
+        Arc<BandwidthSinks>,
+    ) {
```
I think one important property of builders is that you can chain them. Can we store the `BandwidthSinks` inside the phase and return a tuple in the final `build` method?

mxinden replied:
I don't know a way how we can do this without introducing significant complexity into all subsequent phases.

`build` would need to be generic over the fact that the user called `with_bandwidth_logging` in previous steps.

I find it actually quite clean that `with_bandwidth_logging` returns both. I don't find it intuitive if I would get the `BandwidthSinks` as a final step. At that point I would have already forgotten what I am getting.

thomaseizinger replied:
At the moment, where we insert this matters, right?

Could we change the API such that the user has to pass in `Arc<BandwidthSinks>`? Then we could still chain the builder.

mxinden replied:
We could do that. We would only need to make the two constructors public:

- https://github.com/libp2p/rust-libp2p/blob/master/libp2p/src/bandwidth.rs#L49-L54

- https://github.com/libp2p/rust-libp2p/blob/2c3d6ad6da8663f0bc5751138aa163988fadd8ab/libp2p/src/bandwidth.rs#L109-L116

Fine by me.

mxinden replied:
Resolving here. @thomaseizinger please reopen in case you feel strongly about it. We can still do this in the future introducing a new method and deprecating the old.

thomaseizinger commented on libp2p/src/builder.rs:
```diff
@@ -0,0 +1,1795 @@
+    type U = T::ListenerUpgrade;
+}
+
+// TODO: Seal this.
```
Let's not forget about this.

thomaseizinger commented on libp2p/src/builder.rs:
```diff
@@ -0,0 +1,1795 @@
+where
+    B: NetworkBehaviour,
+{
+    // TODO mxinden: why do we need an io error here? isn't box<dyn> enough?
```
I don't remember why.

mxinden replied:
Addressed in b135900b7.

thomaseizinger commented on libp2p/src/builder.rs:
```diff
@@ -0,0 +1,1795 @@
+}
+
+macro_rules! impl_tcp_noise_builder {
+    ($providerKebabCase:literal, $providerCamelCase:ident, $tcp:ident) => {
```
```suggestion
    ($providerKebabCase:literal, $providerPascalCase:ident, $tcp:ident) => {
```

`camelCase` != `PascalCase`.

mxinden replied:
83e3674dc

thomaseizinger commented on libp2p/src/builder.rs:
```diff
@@ -0,0 +1,1795 @@
+            .with_websocket()
+            .with_tls()
+            .unwrap()
+            .with_noise()
```
Hmm, I would have expected this to work differently. Websocket is a base transport (i.e. single stream). I would have expected that I have to specify TLS and/or noise only once for these base transports.

Maybe we need to roll these things up from the other end? Like, start with the "wrapper" transports like DNS , relay and websocket and only specify what they run on last?

mxinden replied:
Agreed that that would be the cleanest solution, but as I argue in https://github.com/libp2p/rust-libp2p/pull/4120#discussion_r1318320011 I don't think this is the easiest solution for users.

Right now each transport is build consecutively. E.g. after you called `with_tcp().with_tls().with_noise()` you are done with TCP, no later method calls will alter the TCP transport to be built.

If we follow your approach above instead, users will need to understand the difference between base transports and full transports. That includes e.g. the complexity of both TCP and relay being a base transport, while QUIC is not. I don't think we should push this complexity onto the user.

What do you think @thomaseizinger with this perspective?

thomaseizinger replied:
This is outdated as of our sync discussion on this. I think we should try the approach of passing sec & muxer protocols as parameters.

mxinden replied:
> I think we should try the approach of passing sec & muxer protocols as parameters.

I don't think we can do that @thomaseizinger. See https://github.com/libp2p/rust-libp2p/pull/4120#issuecomment-1709885968

thomaseizinger reviewed (commented):
Thank you for pushing this forward! It is a great improvement. I've left lots of comments, not all of them are critical :)

mxinden committed:
Update libp2p/CHANGELOG.md

Co-authored-by: Thomas Eizinger <thomas@eizinger.io>

mxinden committed:
Update libp2p/src/lib.rs

Co-authored-by: Thomas Eizinger <thomas@eizinger.io>

mxinden committed:
[WIP] with_tcp including sec and mux

It compiles :tada:

mxinden committed:
[WIP] add support for tls+noise and yamux+mplex

mxinden committed:
Be generic over provider

mxinden commented on libp2p/src/builder.rs:
```diff
@@ -0,0 +1,1741 @@
+                <MuxUpgrade::Upgrade as OutboundUpgrade<Negotiated<SecStream>>>::Future: Send,
+                MuxError: std::error::Error + Send + Sync + 'static,
+                <<<MuxUpgrade as IntoMultiplexerUpgrade<SecStream>>::Upgrade as UpgradeInfo>::InfoIter as IntoIterator>::IntoIter: Send,
+                <<MuxUpgrade as IntoMultiplexerUpgrade<SecStream>>::Upgrade as UpgradeInfo>::Info: Send,
```
Being generic over the security and multiplexer protocol(s) requires a lot of type bounds.

thomaseizinger replied:
Damn, what a wall of text :see_no_evil:

mxinden commented on libp2p/src/builder.rs:
```diff
@@ -0,0 +1,1741 @@
+                // TODO: The single tuple is not intuitive.
+                (libp2p_yamux::Config::default(),),
+            )
+            .unwrap()
```
I don't think the single tuple, specifically the `,`, is intuitive.

mxinden commented on libp2p/src/builder.rs:
```diff
@@ -0,0 +1,1741 @@
+                    libp2p_mplex::MplexConfig::default(),
+                ),
+            )
+            .unwrap()
```
Overall, when used with multiple security protocols and muxers, this is still quite readable.

mxinden commented on libp2p/src/builder.rs:
```diff
@@ -0,0 +1,1741 @@
+    fn into_security_upgrade(self, keypair: &Keypair) -> Result<Self::Upgrade, Self::Error> {
+        self.0(keypair)
+    }
+}
```
Note that one can not implement `IntoSecurityUpgrade` for `F` directly, as such implementation would conflict with the implementation on `(F1, F2)`.

mxinden reviewed (commented):
I followed through with the latest suggestion, namely to have `with_tcp` accept the security and multiplexer protocol(s).

It introduces significant complexity. Both for the user (single element tuple) and the implementation (type bounds). That said, I am fine moving forward with it, especially given that it allows for a clean linear call chain of `with_*` without the unnested `with_noise` and `with_tls`.

@thomaseizinger what do you think? If you agree with the approach as is, I can apply it to `with_relay` and `with_websocket`.

On the meta level, I have spend significant time on this pull request at this point. I don't think the value of the feature would justify spending much more time. I think that time is spend better elsewhere. At this point I would rather like to ship something instead of trying to achieve perfection.

mxinden committed:
Simplify tls and noise construction

thomaseizinger committed:
Remove 1-tuple from security upgrade

thomaseizinger committed:
Resolve single tuple problem for muxer upgrade

thomaseizinger committed:
Undo accidental change

thomaseizinger committed:
Merge branch 'master' into transport-builder

thomaseizinger committed:
Resolve some more warnings

thomaseizinger committed:
Undo unnecessary changes

mxinden committed:
Merge branch 'master' of https://github.com/libp2p/rust-libp2p into transport-builder

mxinden committed:
Remove unnecessary trait bounds on IntoSecurityUpgrade::Error

mxinden committed:
Adjust with_relay

mxinden committed:
Adjust with_websocket

mxinden committed:
Split builder.rs into separate files

mxinden committed:
Update upnp and browser webrtc

mxinden committed:
Rename SwarmConfig to Config

mxinden committed:
Fix wasm compilation

mxinden committed:
More wasm compilation flags

mxinden committed:
Fix ipfs-kad example

mxinden committed:
Remove reference to keep alive behaviour

mxinden committed:
Fix example warning

mxinden committed:
cargo fmt

mxinden committed:
Fix doc tests

mxinden committed:
fix swarm test

mxinden committed:
Fix kad tests

mxinden committed:
Fix swarm config method names

mxinden commented on libp2p/src/builder/phase/other_transport.rs:
```diff
@@ -0,0 +1,188 @@
+    type Error;
+
+    fn try_into_transport(self) -> Result<T, Self::Error>;
+}
```
Haven't figures out how to seal this trait yet. If I implement `Sealed` for both `T` and `Result<T, Box<dyn std::error::Error + Send + Sync>>` I get a conflicting implementations error.

mxinden replied:
Fixed with 52ac76cdb129fdcc861fb48e3a74ac919f1951c9

mxinden committed:
Rename to with_relay_client

mxinden committed:
Implement with_websocket for RelayPhase

mxinden committed:
Rename to IdentityPhase

mxinden committed:
Introduce BehaviourError

mxinden committed:
Introduce TransportError

mxinden committed:
Refactor nested quic macro

mxinden committed:
Reword note on dummy transport usage

mxinden committed:
Rephase note on unnecessary tokio await

mxinden committed:
Seal TryInto traits

mxinden committed:
Merge branch 'master' of https://github.com/libp2p/rust-libp2p into transport-builder

mxinden committed:
Rename to providerPascalCase

mxinden committed:
interop: re-introduce mplex

mxinden committed:
Refactor websocket error handling

mxinden committed:
Remove without_executor

mxinden committed:
Rename new_with_config to new

mxinden committed:
Reword libp2p changelog

mxinden committed:
Reword swarm changelog

mxinden committed:
refactor(examples/dcutr): use tokio

mxinden committed:
Minor fixes

mxinden commented on swarm/src/connection.rs:
```diff
@@ -757,7 +759,7 @@ mod tests {
                     counter: alive_substream_counter.clone(),
                 }),
-                MockConnectionHandler::new(Duration::ZERO),
+                MockConnectionHandler::new(Duration::from_secs(10)),
```
Without it, the upgrade instantly times out and the limit can never be reached.

thomaseizinger commented on examples/browser-webrtc/src/main.rs:
```diff
@@ -28,19 +27,22 @@ async fn main() -> anyhow::Result<()> {
+        })?
+        .with_behaviour(|_| ping::Behaviour::default())?
+        .with_swarm_config(
+            libp2p::swarm::Config::with_tokio_executor().with_idle_connection_timeout(
```
Should we have `with_swarm_config` take a closure that already hands you to the default instance of `swarm::Config`? I am worried that users need to now align their usage of `with_tokio_executor` here with the above `with_tokio`.

mxinden replied:
Good idea @thomaseizinger. Done in 1fd5a5f6d.

thomaseizinger commented on examples/identify/src/main.rs:
```diff
@@ -20,35 +20,31 @@
+                key.public(),
+            ))
+        })
+        .unwrap()
```
```suggestion
        .with_behaviour(|key| {
            identify::Behaviour::new(identify::Config::new(
                "/ipfs/id/1.0.0".to_string(),
                key.public(),
            ))
        })?
```

mxinden replied:
7a8ee9b1d

thomaseizinger commented on examples/identify/src/main.rs:
```diff
@@ -20,35 +20,31 @@
-
-    let mut swarm =
-        SwarmBuilder::with_async_std_executor(transport, behaviour, local_peer_id).build();
+    println!("Local peer id: {:?}", swarm.local_peer_id());
```
We removed these in a previous PR.

mxinden replied:
e962f0297

thomaseizinger commented on examples/ipfs-kad/src/main.rs:
```diff
@@ -41,30 +36,33 @@ const BOOTNODES: [&str; 4] = [
+            let store = kad::store::MemoryStore::new(key.public().to_peer_id());
+            kad::Behaviour::with_config(key.public().to_peer_id(), store, cfg)
+        })
+        .unwrap()
```
```suggestion
        })?
```

mxinden replied:
358e7ff67

thomaseizinger commented on examples/ping/src/main.rs:
```diff
@@ -21,31 +21,21 @@
+            yamux::Config::default,
+        )?
+        .with_behaviour(|_| ping::Behaviour::default())
+        .unwrap()
```
```suggestion
        .with_behaviour(|_| ping::Behaviour::default())?
```

mxinden replied:
711d7b6c7

thomaseizinger commented on examples/ping/src/main.rs:
```diff
@@ -21,31 +21,21 @@
-
-    let mut swarm =
-        SwarmBuilder::with_async_std_executor(transport, ping::Behaviour::default(), local_peer_id)
-            .idle_connection_timeout(Duration::from_secs(60)) // For illustrative purposes, keep idle connections alive for a minute so we can observe a few pings.
```
Missing idle connection timeout in the new version.

mxinden replied:
99d7b6dc2

thomaseizinger commented on interop-tests/Cargo.toml:
```diff
@@ -20,8 +20,11 @@ rand = "0.8.5"
+libp2p-noise = { workspace = true }
+libp2p-tls = { workspace = true }
+libp2p-webrtc = { workspace = true, features = ["tokio"] }
+libp2p-yamux = { workspace = true }
```
Why do we need this?

mxinden replied:
Good catch. Thanks. 9bd04c73c

thomaseizinger commented on interop-tests/src/arch.rs:
```diff
@@ -47,87 +36,131 @@ pub(crate) mod native {
+            (Transport::Ws, Ok(SecProtocol::Tls), Ok(Muxer::Yamux)) => {
+                let swarm = libp2p::SwarmBuilder::with_new_identity()
+                    .with_tokio()
+                    .with_websocket(libp2p_tls::Config::new, yamux::Config::default)
```
Why can't this just be called via the module that is imported?

```suggestion
                    .with_websocket(tls::Config::new, yamux::Config::default)
```

mxinden replied:
:+1: 5a8c85393

thomaseizinger commented on interop-tests/src/arch.rs:
```diff
@@ -177,35 +209,40 @@ pub(crate) mod wasm {
+                        libp2p::webtransport_websys::Transport::new(
+                            libp2p::webtransport_websys::Config::new(key),
+                        )
+                        .map(|(peer_id, conn), _| (peer_id, StreamMuxerBox::new(conn)))
```
Should we do this `.map` for the user inside the builder?

mxinden replied:
I am hesitant to do so. What if the transport already returns a `StreamMuxerBox`?

thomaseizinger replied:
> I am hesitant to do so. What if the transport already returns a `StreamMuxerBox`?

Are you worried about an additional allocation? Or the indirection caused by another pointer look-up?

I think this would be a nice usability improvement. It is a problem that users are facing, for example: https://github.com/libp2p/rust-libp2p/issues/3179

This is only present for obscure transport combinations which usually only happens with alpha-releases of our transports. Unless there are concerns other than the above performance implications, I'd say that the usability is more important.

thomaseizinger replied:
It would be great if we could tag this PR as resolving https://github.com/libp2p/rust-libp2p/issues/3179 :)

thomaseizinger commented on libp2p/src/builder/phase/behaviour.rs:
```diff
@@ -0,0 +1,90 @@
+
+#[derive(Debug, thiserror::Error)]
+#[error("failed to build behaviour: {0}")]
+pub struct BehaviourError(Box<dyn std::error::Error + Send + Sync>);
```
I think we want `'static` here otherwise the downcasting APIs don't work on it, right?

mxinden replied:
371b380ca :+1:

thomaseizinger commented on libp2p/src/builder/phase/build.rs:
```diff
@@ -0,0 +1,29 @@
+    pub(crate) swarm_config: libp2p_swarm::Config,
+}
+
+const CONNECTION_TIMEOUT: std::time::Duration = std::time::Duration::from_secs(10);
```
This is currently not configurable, right? Should we add it to `swarm::Config`?

mxinden replied:
I would do so once we have a user for it. I would guess that `10s` covers 99% of all cases. Can you think of a use-case that would require the config option today already?

thomaseizinger replied:
Fair enough!

thomaseizinger commented on libp2p/src/builder/phase/swarm.rs:
```diff
@@ -0,0 +1,80 @@
+}
+
+impl<T, B, Provider> SwarmBuilder<Provider, SwarmPhase<T, B>> {
+    pub fn with_swarm_config(
```
I don't feel strongly but do we need the `swarm` mention here? This function is defined on a struct called `SwarmBuilder`. Isn't that enough context?

mxinden replied:
With all the other type above, e.g. tcp, quic, relay, behaviour, I find it more intuitive to stress that this is the config of the swarm now. Thus I prefer `with_swarm_config` instead of `with_config`.

thomaseizinger commented on libp2p/src/builder/select_security.rs:
```diff
@@ -0,0 +1,112 @@
+// Copyright 2018 Parity Technologies (UK) Ltd.
```
```suggestion
// Copyright 2023 Protocol Labs.
```

mxinden replied:
a5dc83c1e

thomaseizinger commented on libp2p/src/tutorials/ping.rs:
```diff
@@ -71,50 +71,39 @@
+//!     let mut swarm = libp2p::SwarmBuilder::with_new_identity()
+//!         .with_async_std()
+//!         .with_tcp(
+//!             Default::default(),
```
Let's not use `Default::default` here please.

mxinden replied:
8c12ec51b

thomaseizinger commented on misc/server/src/main.rs:
```diff
@@ -77,35 +71,25 @@ async fn main() -> Result<(), Box<dyn Error>> {
+            noise::Config::new,
+            yamux::Config::default,
+        )?
+        .with_quic_config(|key| {
```
Would `with_custom_quic` perhaps be a better name? At the moment we have `with_quic` and `with_quic_config` and `with_quic_config` implies `with_quic` but I think technically, that it is not super obvious.

Also, we could change this closure to already give you an instance of `quic::Config`! Then the user would a) not have to name it and b) for builder-style configs, they can do them as one-liners:

```rust
	.with_quic_config(|cfg| cfg.with_draft_29().with_xyz())
```

mxinden replied:
I find `with_quic_config` more intuitive. While `with_quic_config` implies that one can configure a `libp2p_quic::Config`, `with_custom_quic` does not imply what one can customize.

Do you worry that users think that `with_quic_config` only constructs the `libp2p_quic::Config` but not the actual `Transport`?

4743616ec makes `with_quic_config` take a closure.

Note that `libp2p_quic::Config` does not provide methods that can be chained. In case that is something worth doing, we can still do so in a follow-up pull request.

thomaseizinger replied:
> Do you worry that users think that `with_quic_config` only constructs the `libp2p_quic::Config` but not the actual `Transport`?

Not really, because there is no alternative function. I just figured the naming could be perceived as being a bit off :)

thomaseizinger commented on misc/server/src/main.rs:
```diff
@@ -139,10 +122,11 @@ async fn main() -> Result<(), Box<dyn Error>> {
+        block_on(http_service::metrics_server(
+            metric_registry,
+            opt.metrics_path,
+        ))
```
Is this a bad merge or did you change that back on purpose?

thomaseizinger reviewed (commented):
Great work! I went through all the _usage_ sites and it already looks super clean! I've made a few more suggestions that could optimise ergonomics a bit more.

Thank you for pushing this forward!

mxinden committed:
Have with_swarm_config take closure

mxinden committed:
Have with_quic_config take closure

mxinden committed:
Remove unwrap

mxinden committed:
Remove duplicate local peer id print

mxinden committed:
Handle behaviour error

mxinden committed:
Handle behaviour error

mxinden committed:
Ping idle connection timeout

mxinden committed:
Interop remove separate yamux import

mxinden committed:
Use tls instead of libp2p-tls

mxinden committed:
Require errors to be 'static

mxinden committed:
Extend license header

mxinden committed:
Use libp2p_tcp::Config::default()

mxinden committed:
Fix bad merge

mxinden committed:
Reduce diff

mxinden committed:
Merge branch 'master' of https://github.com/libp2p/rust-libp2p into transport-builder

mxinden committed:
Use new SwarmBuilder in examples/browser-webrtc wasm

mxinden committed:
fmt

thomaseizinger commented on examples/browser-webrtc/src/lib.rs:
```diff
@@ -18,19 +18,17 @@ pub async fn run(libp2p_endpoint: String) -> Result<(), JsError> {
+        .with_wasm_bindgen()
+        .with_other_transport(|key| {
+            libp2p_webrtc_websys::Transport::new(libp2p_webrtc_websys::Config::new(&key))
+                .map(|(peer_id, conn), _| (peer_id, StreamMuxerBox::new(conn)))
```
Nit: I know you've just replicated what was there before but could we follow our convention of importing based on the module here?

thomaseizinger commented on examples/metrics/src/main.rs:
```diff
@@ -39,21 +37,16 @@ mod http_service;
+            yamux::Config::default,
+        )?
+        .with_behaviour(|key| Behaviour::new(key.public()))?
+        .with_swarm_config(|cfg| cfg.with_idle_connection_timeout(Duration::from_secs(u64::MAX)))
```
Nice, I am glad the formatter now puts this on a single line.

thomaseizinger commented on interop-tests/src/arch.rs:
```diff
@@ -47,87 +36,131 @@ pub(crate) mod native {
+            (Transport::Ws, Ok(SecProtocol::Tls), Ok(Muxer::Yamux)) => {
+                let swarm = libp2p::SwarmBuilder::with_new_identity()
+                    .with_tokio()
+                    .with_websocket(tls::Config::new, yamux::Config::default)
```
Our `libp2p-websocket` crate doesn't yet follow the structure of `Config` and `Transport`. I'd imagine that we will have to change this at some point to also take a `websocket::Config` as the first parameter similar to how we do TCP. Otherwise, `/wss` is not supported using this builder.

mxinden replied:
Yes. We can do so in the future by introducing a `with_websocket_config`.

thomaseizinger replied:
The more consistent approach would be to add another argument like we did for TCP.

thomaseizinger commented on libp2p/src/tutorials/ping.rs:
```diff
@@ -307,16 +293,14 @@
-//!     let behaviour = ping::Behaviour::default();
-//!
-//!     let mut swarm = SwarmBuilder::with_async_std_executor(transport, behaviour, local_peer_id)
-//!         .idle_connection_timeout(Duration::from_secs(30)) // Allows us to observe pings for 30 seconds.
```
This tutorial is missing the idle connection timeout.

mxinden replied:
Good catch. 3717d1fe2.

thomaseizinger commented on libp2p/src/tutorials/ping.rs:
```diff
@@ -149,26 +138,25 @@
-//! [`NetworkBehaviour`] to the [`Transport`] as well as events from the
-//! [`Transport`] to the [`NetworkBehaviour`]. As you can see, after [`Swarm`] initialization, we
+//! Now that we have a [`Transport`] and a [`NetworkBehaviour`], we can build the [`Swarm`]
+//! whcihconnects the two, allowing both to make progress. Put simply, a [`Swarm`] drives both a
```
```suggestion
//! which connects the two, allowing both to make progress. Put simply, a [`Swarm`] drives both a
```

thomaseizinger commented on libp2p/src/tutorials/ping.rs:
```diff
@@ -181,15 +169,15 @@
 //!
 //! #[async_std::main]
 //! async fn main() -> Result<(), Box<dyn Error>> {
-//!     env_logger::init();
```
Logger is gone now but still mentioned in the tutorial.

mxinden replied:
3717d1fe2

thomaseizinger commented on misc/server/src/main.rs:
```diff
@@ -77,35 +68,24 @@ async fn main() -> Result<(), Box<dyn Error>> {
-        .map(|either_output, _| match either_output {
-            Either::Left((peer_id, muxer)) => (peer_id, StreamMuxerBox::new(muxer)),
-            Either::Right((peer_id, muxer)) => (peer_id, StreamMuxerBox::new(muxer)),
+    info!("Local peer id: {local_peer_id}");
```
Bad merge.

mxinden replied:
Thanks bb91e8ba2

thomaseizinger reviewed (commented):
Good work @mxinden! Some final comments :)

mxinden committed:
Use webrtc_websys module

mxinden committed:
Use with_tokio_executor

mxinden committed:
Update libp2p/src/tutorials/ping.rs

Co-authored-by: Thomas Eizinger <thomas@eizinger.io>

mxinden committed:
Re-introduce env_logger and fix idle timeout

mxinden committed:
Remove duplicate peer id printing

thomaseizinger reviewed (approved):
Super clean! Very excited to land this :)

I think the cherry on the cake would be if we also remove the `.map` to `StreamMuxerBox` but if you feel strongly, happy to land this as is.

mxinden committed:
Don't require StreamMuxerBox in with_other_transport

mxinden committed:
fmt

mxinden committed:
Merge branch 'master' of https://github.com/libp2p/rust-libp2p into transport-builder

mxinden committed:
Minor follow-ups

mxinden committed:
Fix wasm arch

thomaseizinger commented on examples/dcutr/src/main.rs:
```diff
@@ -117,21 +83,29 @@ fn main() -> Result<(), Box<dyn Error>> {
+            })?
+            .build();
+
+    info!("Local peer id: {:?}", swarm.local_peer_id());
```
```suggestion
```

thomaseizinger commented on examples/rendezvous/src/bin/rzv-discover.rs:
```diff
@@ -20,41 +20,39 @@
-    .idle_connection_timeout(Duration::from_secs(5))
-    .build();
+        })
+        .unwrap()
```
```suggestion
        })?
```

thomaseizinger commented on libp2p/src/builder/phase/tcp.rs:
```diff
@@ -0,0 +1,208 @@
+                // Note that this allows a user to build a faulty (i.e. dummy) transport only.
+                // Ideally the final build step would not exist unless one adds another transport
+                // after `without_tcp`. Though currently this would introduce significant
+                // complexity to each following step.
```
Given that this function is crate-private, does this comment still hold value?

mxinden replied:
Good point. It is used in the various shortcut methods. Though those themselves enforce the user to provide a valid transport.

thomaseizinger commented on libp2p/src/builder/phase/tcp.rs:
```diff
@@ -0,0 +1,208 @@
+            feature = $providerKebabCase,
+        ))]
+        impl SwarmBuilder<$providerPascalCase, TcpPhase> {
+            pub fn with_tcp<SecUpgrade, SecStream, SecError, MuxUpgrade, MuxStream, MuxError>(
```
I'd like us to have some basic docs on here that explain to the user that `security_upgrade` and `multiplexer_upgrade` are function-pointers!

mxinden replied:
:+1: e0029e485

thomaseizinger reviewed (approved):
This is the best thing since sliced-bread that happened to libp2p in regards to UX! Looking forward to shipping this :)

thomaseizinger committed:
Merge branch 'master' into transport-builder

mxinden committed:
Use #[allow(unused_imports)]

mxinden committed:
Remove obsolete peer id print

mxinden committed:
Don't unwrap

mxinden committed:
Remove obsolete comment

mxinden committed:
Document with_tcp with_relay_client with_websocket taking function pointer

mxinden committed:
Merge branch 'master' of https://github.com/libp2p/rust-libp2p into transport-builder

mxinden committed:
Fix intra-doc link

thomaseizinger committed:
Merge branch 'master' into transport-builder

thomaseizinger committed:
Merge branch 'master' into transport-builder

thomaseizinger committed:
Reduce diff

thomaseizinger committed:
Reduce diff

thomaseizinger committed:
Fix bug sec protocol selection

thomaseizinger committed:
Configure idle-connection timeout

thomaseizinger committed:
Merge branch 'master' into transport-builder

thomaseizinger committed:
Fix interop config bug

mxinden commented on interop-tests/src/arch.rs:
```diff
@@ -120,7 +120,7 @@ pub(crate) mod native {
                 libp2p::SwarmBuilder::with_new_identity()
                     .with_tokio()
-                    .with_websocket(tls::Config::new, mplex::MplexConfig::default)
+                    .with_websocket(noise::Config::new, mplex::MplexConfig::default)
```
Once more, great catch @thomaseizinger.

thomaseizinger replied:
Interop tests made this one easy :)
